src/main/java/bio/terra/janitor/service/cleanup/FlightManager.java

```diff
@@ -91,17 +92,27 @@ public FlightManager(
    */
   private Optional<TrackedResource> updateResourceForCleaning(
       Instant expiredBy, String flightId, TransactionStatus unused) {
-    Optional<TrackedResource> resource =
-        janitorDao.retrieveExpiredResourceWith(expiredBy, TrackedResourceState.READY);
-    if (!resource.isPresent()) {
+    List<TrackedResource> resources =
+        janitorDao.retrieveResourcesMatching(
+            TrackedResourceFilter.builder()
+                .allowedStates(ImmutableSet.of(TrackedResourceState.READY))
+                .expiredBy(expiredBy)
+                .limit(1)
+                .build());
+    if (resources.isEmpty()) {
       return Optional.empty();
     }
+    if (resources.size() > 1) {
```

Comment:
I think you have `limit(1)` in the filter, and it will never reach here. Maybe remove the limit?

Reply:
I want the limit so that this only processes one resource at a time. At the same time, I have this check to make sure that it's behaving as we expect, otherwise there's a bug and something is broken. Easier to fail early.